sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -4249,7 +4250,30 @@ object ApplyCharTypePadding extends Rule[LogicalPlan] {
  * rule right after the main resolution batch.
  */
 object RemoveTempResolvedColumn extends Rule[LogicalPlan] {
-  override def apply(plan: LogicalPlan): LogicalPlan = plan.resolveExpressions {
-    case t: TempResolvedColumn => UnresolvedAttribute(t.nameParts)
+  override def apply(plan: LogicalPlan): LogicalPlan = {
+    plan.foreachUp {
+      // Having clause could be resolved as a Filter. When having func(column with wrong data type),
+      // the column could be wrapped by a TempResolvedColumn, e.g. mean(tempresolvedcolumn(t.c)).
+      // Because TempResolvedColumn can still preserve column data type, here is a chance to check
+      // if the data type matches with the required data type of the function. We can throw an error
+      // when data types mismatches.
+      case operator: Filter =>
+        operator.expressions.foreach(_.foreachUp {
+          case e: Expression if e.checkInputDataTypes().isFailure =>
```

Comment:
@cloud-fan 

I updated this PR based on our discussion.

Right now my implementation could fail a few queries (for example SQLQuerySuite.normalize special floating numbers in subquery). 

```
Invalid call to dataType on unresolved object
org.apache.spark.sql.catalyst.analysis.UnresolvedException: Invalid call to dataType on unresolved object
	at org.apache.spark.sql.catalyst.analysis.UnresolvedAttribute.dataType(unresolved.scala:137)
	at org.apache.spark.sql.catalyst.expressions.BinaryOperator.checkInputDataTypes(Expression.scala:713)
	at org.apache.spark.sql.catalyst.expressions.BinaryComparison.checkInputDataTypes(predicates.scala:908)
	at org.apache.spark.sql.catalyst.analysis.RemoveTempResolvedColumn$.$anonfun$apply$82(Analyzer.scala:4262)

```

I believe it is because I called `checkInputDataTypes()` on a binary comparator that has at least one side as `UnresolvedAttribute`. Do you know how to filter such case in my implementation? 

I am not familiar with scala and our codebase. Any help will be much appreciated.

Reply:
I tried those functions like `transformWithPruning`, unfortunately didn't make it work.

Reply:
A straightforward fix is to change it to `case e: Expression if e.childrenResolved && e.checkInputDataTypes().isFailure`.

A new idea is to not strip `TempResolvedColumn`, as `TempResolvedColumn` always means a failure. We can update `CheckAnalysis` to handle it, i.e. adding a new case after `case e: Expression if e.checkInputDataTypes().isFailure =>`
```
case t: TempResolvedColumn =>
  val a = UnresolvedAttribute(t.nameParts)
  the same code that handles "case a: Attribute if !a.resolved"
```

Reply:
@cloud-fan 

thank you! `e.childrenResolved` is a handy call and it indeeds solves problem!

I am still checking error at `RemoveTempResolvedColumn`. If you actually prefer to check it at `CheckAnalysis`, let me know and I can make a change.

Reply:
In fact I am not sure the consequence of not striping `TempResolvedColumn`. I would guess there was a reason to add that `RemoveTempResolvedColumn` rule/batch? Otherwise why not deal with `TempResolvedColumn` in `CheckAnalysis` when `TempResolvedColumn` was introduced?